FCMExternalProberTarget/app/src/main/java/com/google/firebase/messaging/testing/fcmexternalprobertarget/FCMReceiveService.java

```
@@ -0,0 +1,109 @@
+    private Clock logTimer;
+
+    /**
+     * Create an instance for testing
+     * @param context Mocked Context object
+     * @param viewLogging Whether logs should be written to UI
+     */
+    public FCMReceiveService (Context context, boolean viewLogging, Clock logTimer) {
+        this.context = context;
+        this.viewLogging = viewLogging;
+        this.logTimer = logTimer;
+    }
+
+    public FCMReceiveService () {
+        this.context = this;
+        viewLogging = false;
+        logTimer = Clock.systemUTC();
+    }
+
+    @Override
+    public void onNewToken(@NonNull String token) {
+        // Store token in a file on external storage so it can be accessed by the probe
+        File tokenFile = makeExternalFile("","token.txt");
+        writeToFile(tokenFile, token);
+        logToUI("Info", "Token generated");
+    }
+
+    @Override
+    public void onMessageReceived(@NonNull RemoteMessage remoteMessage) {
+        long receivedTime = logTimer.instant().getEpochSecond();
+        String sendTime = remoteMessage.getData().get("sendTime");
+        File logFile = makeExternalFile("logs", sendTime + ".txt");
+        writeToFile(logFile, Long.toString(receivedTime));
+        logToUI("Info","Message Received: " + sendTime + ".txt");
+    }
+
+
+    private void logToUI(String tag, String logText) {
+        Log.d(tag, logText);
+        if (viewLogging) {
+            return;
+        }
+        Intent intent = new Intent("updateUI");
+        intent.putExtra("logText", tag + ": " + logText);
+        LocalBroadcastManager.getInstance(this).sendBroadcast(intent);
+    }
+
+    private File makeExternalFile(String path, String fileName) {
+        File newPath = new File(context.getExternalFilesDir(null), path);
+        if (!newPath.exists()) {
+            if (!newPath.mkdirs()) {
+                logToUI("Error", "Unable to create directory " + path);
+            }
+            logToUI("Info", "Directory " + path + " created");
```

This is a silent failure, since it always returns a new file whether or not it works.  

Since writeToFile also silently fails this is probably fine in this case, and other code isn't going to be built directly on top of this.  But generally this is a pattern to avoid.

---

addressed in f88a918